shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/LedgerState.hs

```diff
@@ -801,20 +807,30 @@ witsVKeyNeeded ::
   GenDelegs crypto ->
   WitHashes crypto
 witsVKeyNeeded utxo' tx@(Tx txbody _ _) genDelegs =
-  WitHashes $
-    certAuthors
-      `Set.union` inputAuthors
-      `Set.union` owners
-      `Set.union` wdrlAuthors
-      `Set.union` updateKeys
+  WitHashes
+    { bootstrapWitHashes = bootstrapInputAuthors,
+      properWitHashes =
+        certAuthors
+          `Set.union` inputAuthors
+          `Set.union` owners
+          `Set.union` wdrlAuthors
+          `Set.union` updateKeys
+    }
   where
     inputAuthors :: Set (KeyHash 'Witness crypto)
-    inputAuthors = foldr accum Set.empty (_inputs txbody)
+    bootstrapInputAuthors :: Set (BootstrapKeyHash crypto)
+    (inputAuthors, bootstrapInputAuthors) =
+      foldr accum (Set.empty, Set.empty) (_inputs txbody)
       where
-        accum txin ans =
+        accum txin ans@(whs, bootstrapWhs) =
```

Comment:
Could use a `WitHashes` as the accumulator, if you think that'd be clearer than the tuple and bang patterns?

Reply:
I think it would be.